ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/FinalProjectNode.java

```diff
@@ -114,10 +119,28 @@ public void validateKeyPresent(final SourceName sinkName) {
             return true;
           }
 
-          return parentSchema.isKeyColumn(columnName);
+          if (parentSchema.isKeyColumn(columnName)) {
+            seenKeyColumns.computeIfAbsent(columnName, k -> new HashSet<>()).add(se.getAlias());
+            return true;
+          }
         }
         return false;
       });
+
+      for (final Entry<ColumnName, Set<ColumnName>> seenKey : seenKeyColumns.entrySet()) {
+        if (seenKey.getValue().size() > 1) {
+          final String keys = GrammaticalJoiner.and().join(
+              seenKey.getValue().stream().map(Name::text).sorted());
+          throw new KsqlException("The projection contains a key column (" + seenKey.getKey()
```

Comment:
Not directly related to your PR, but can you help me understand why the validation that key columns are not selected more than once is performed in `FinalProjectNode` while the validation that key columns are selected is performed in `DataSourceNode`, rather than performing both checks in the same place?

Reply:
Every node "validates" that the key is selected, they just all delegate to the child node that has that information:
```java
// PlanNode.java:L145
  void validateKeyPresent(final SourceName sinkName, final Projection projection) {
    getSources().forEach(s -> s.validateKeyPresent(sinkName, projection));
  }
```

`DataSourceNode`, `JoinNode` and `UserRepartitionNode` all override this method because those are the nodes that truly "know" whether the key was selected. Imagine we projected the key column with a different name, the schema in the `FinalProjectNode` would consider that the key - but of course we didn't select that! We selected the original key. 

To illustrate, in the example below the project node has a schema `[col1 INT PRIMARY KEY]` but we didn't select `col1` - so it would fail! Of course we could do some magic and "remember" that `b.col1` is actually `a.id`, but why not just delegate that to the node that already tracks that?
```sql
CREATE TABLE a (id INT PRIMARY KEY, col1 INT);
CREATE TABLE b AS SELECT id AS col1; 
```

(Note that it's a little confusing, because there is also `VerifiableNode`, which `AggregateNode`, `FinalProjectNode` and `SuppressNode` all override - and it is `VerifiableNode#validateKeyPresent` is what is called at the top level, but these just delegate down to the three source nodes eventually). 

> why the validation that key columns are not selected more than once is performed in `FinalProjectNode`

I think this one is clearer why it can't be in `DataSourceNode`, it's the other way around that might need some justification.

---

All that being said, I think this can definitely be made clearer going forward. The logical nodes don't follow a very strict abstraction model unfortunately.

Reply:
Thanks for the explanation -- super helpful!